### ISSUE TITLE
Question about disc_mutual_info_loss

### ISSUE THREAD

**Issue by oliver-snowdon** [EmilienDupont/infogan#2]
Hi Emilien,

I'm trying to understand how an infoGAN works, and I was looking at your code as an example. On line 18 in util.py, it appears that the "ent" tensor will not have any gradient with respect to the weights of the network, because it only depends on y_true, so the network should behave the same in training without it. Is this correct? Also, are you sure that on line 19 you want to subtract cross_ent, not add it? Surely if you subtract it, then as the value corresponding to the "one hot" label in y_true increases (as you want it to do) the log of it will also increase, so the crossentropy will decrease (as you want it to do). However, this will mean the loss function increases, so the opposite will tend to happen.

Cheers,
Oliver

**Comment by EmilienDupont** [comment]
Hi Oliver,

Thanks for your comments and questions and sorry about the slow reply! 

You are correct, the `ent` tensor will not have any gradients with respect to the weights of the network since we assume the true distribution of codes is fixed and so the `ent` term is a constant (see e.g. the top of page 4 in the InfoGAN paper).

And on line 19 you do want to subtract the `cross_ent` term (see equation 4 and 5 in the paper). As you correctly mention, we want the value of `aux_dist` corresponding to the true label in `c_disc` to increase. As this value increases, `cross_ent` decreases and so `disc_mutual_information_loss` increases. However, we are not minimizing `disc_mutual_information_loss`, we are minimizing -1 * `disc_mutual_information_loss` (see line 138 in `infogan.py`), so we end up maximizing `disc_mutual_information_loss` and hence minimizing `cross_ent`, which is the correct behavior.

Hope that makes sense!
